modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/toolbarbuttons/plugin.js

```diff
@@ -50,22 +282,68 @@
 			});
 
 			editor.ui.addRichCombo('TableHeaders', {
+				_headersLabels: {
+					[HEADING_BOTH]: `${editor.lang.table.headers}: ${editor.lang.table.headersBoth}`,
+					[HEADING_COL]: `${editor.lang.table.headers}: ${editor.lang.table.headersColumn}`,
+					[HEADING_NONE]: `${editor.lang.table.headers}: ${editor.lang.table.headersNone}`,
+					[HEADING_ROW]: `${editor.lang.table.headers}: ${editor.lang.table.headersRow}`,
+				},
+
+				_setHeading(heading, editor) {
+					editor.execCommand(`tableHeading${heading}`);
+
+					this.setValue(heading, this._headersLabels[heading]);
+				},
+
 				init() {
-					const headersPrefix = editor.lang.table.headers;
+					this.add(
+						HEADING_NONE,
+						this._headersLabels[HEADING_NONE],
+						this._headersLabels[HEADING_NONE]
+					);
+					this.add(
+						HEADING_ROW,
+						this._headersLabels[HEADING_ROW],
+						this._headersLabels[HEADING_ROW]
+					);
+					this.add(
+						HEADING_COL,
+						this._headersLabels[HEADING_COL],
+						this._headersLabels[HEADING_COL]
+					);
+					this.add(
+						HEADING_BOTH,
+						this._headersLabels[HEADING_BOTH],
+						this._headersLabels[HEADING_BOTH]
+					);
 
-					const headersNone = `${headersPrefix}: ${editor.lang.table.headersNone}`;
-					const headersRow = `${headersPrefix}: ${editor.lang.table.headersRow}`;
-					const headersColumn = `${headersPrefix}: ${editor.lang.table.headersColumn}`;
-					const headersBoth = `${headersPrefix}: ${editor.lang.table.headersBoth}`;
+					const currentHeading = getHeading(null, editor);
 
-					this.add(headersNone, headersNone, headersNone);
-					this.add(headersRow, headersRow, headersRow);
-					this.add(headersColumn, headersColumn, headersColumn);
-					this.add(headersBoth, headersBoth, headersBoth);
+					this._setHeading(currentHeading, editor);
 				},
 
 				label: editor.lang.table.headers,
 
+				onClick(selectedHeading) {
+					this._setHeading(selectedHeading, editor);
+				},
+
+				onRender() {
+					const currentHeading = getHeading(null, editor);
+
+					this._setHeading(currentHeading, editor);
+
+					editor.on(
+						'selectionChange',
+						function (_) {
+							const currentHeading = getHeading(null, editor);
+
+							this._setHeading(currentHeading, editor);
+						},
+						this
+					);
+				},
```

Comment:
Why do we need to update headings on each render, and even more, on each selection? :thinking: 

I would imagine `onClick` is the only event we need.

Reply:
I needed to sync the value of the richCombo even without selecting a value. For example, when we have a published blog article containing a table with already set table heading option and we want to see which is the selected value. This logic is for it. Not just for the first render(see the scope outside selectionChange callback) but for when changing the selection :)

Reply:
For testing, I remember placing log points at each setHeading around the plugin scope to see if left a call that don't cause any effect and I saw everything was being called appropriadely

Reply:
> when we have a published blog article containing a table with already set table heading option and we want to see which is the selected value

If this is the intent, we should trigger toolbar update only once, on initial open. `onRender` and on `'selectionChange'`  are an overkill, triggering many many times in editor use.

Reply:
When removing 'selectionChange' handler:

https://user-images.githubusercontent.com/7663513/155214626-bf49dc50-8ceb-4ad1-ba5e-e3bd53074558.mov

Note "Headers: First Row" is not persisted. So, we should do this overkill 😅